source/common/common/thread.cc

```diff
@@ -1,36 +1,103 @@
 #include "source/common/common/thread.h"
 
+#include <thread>
+
+#include "source/common/common/assert.h"
+#include "source/common/common/macros.h"
+
 namespace Envoy {
 namespace Thread {
 
-bool MainThread::isMainThread() {
-  // If threading is off, only main thread is running.
-  auto main_thread_singleton = MainThreadSingleton::getExisting();
-  if (main_thread_singleton == nullptr) {
-    return true;
+namespace {
+
+// Singleton structure capturing which thread is the main dispatcher thread, and
+// which is the test thread. This info is used for assertions around catching
+// exceptions and accessing data structures which are not mutex-protected, and
+// are expected only from the main thread.
+//
+// TODO(jmarantz): avoid the singleton and instead have this object owned
+// by the ThreadFactory. That will require plumbing the API::API into all
+// call-sites for isMainThread(), which might be a bit of work, but will make
+// tests more hermetic.
+struct ThreadIds {
+  // Determines whether we are currently running on the main-thread or
+  // test-thread. We need to allow for either one because we don't establish
+  // the full threading model in all unit tests.
+  bool inMainOrTestThread() const {
+    std::thread::id id = std::this_thread::get_id();
+    absl::MutexLock lock(&mutex_);
+    return ((main_thread_id_.has_value() && (main_thread_id_.value() == id)) ||
+            (test_thread_id_.has_value() && (test_thread_id_.value() == id)));
+  }
+
+  // Returns a singleton instance of this. The instance is never freed.
+  static ThreadIds& get() { MUTABLE_CONSTRUCT_ON_FIRST_USE(ThreadIds); }
+
+  // Call this when the MainThread exits. Nested semantics are supported, so
+  // that if multiple MainThread instances are declared, we unwind them
+  // properly.
+  void releaseMainThread() {
+    absl::MutexLock lock(&mutex_);
+    ASSERT(std::this_thread::get_id() == main_thread_id_.value());
+    if (--main_thread_use_count_ == 0) {
+      main_thread_id_ = absl::nullopt;
+    }
+  }
+
+  // Call this when the MainThread exits. Nested semantics are supported, so
+  // that if multiple TestThread instances are declared, we unwind them
+  // properly.
+  void releaseTestThread() {
+    absl::MutexLock lock(&mutex_);
+    ASSERT(std::this_thread::get_id() == test_thread_id_.value());
+    if (--test_thread_use_count_ == 0) {
```

Comment:
Can we use a stack of thread id to support nested thread instance? And check the current thread id against the thread id on top the the stacks of main thread ids and test thread ids?

Reply:
we could, but why? it doesn't make sense to nest MainThread or TestThreads -- we have only one of each.

Reply:
To clarify, I need to have nested declarations for MainThread for a few reasons as tests and binaries vary in entry-point. Some MainThread declarations needed for tests, for example, are redundant in production binaries.

I think I actually don't need tested TestThread declarations but it was easy enough to implement them following the pattern for MainThread.

Reply:
I see. Can you elaborate on why it is necessary to keep a reference count to reset the thread id?

Reply:
If we register and release thread id in the same thread in order, is a lock necessary here?